### PR TITLE
twap: Add Multi-Pool hooks test for `AfterPoolCreated`

### PR DESCRIPTION
**Pull request by mattverse**
Sub component of : https://github.com/osmosis-labs/osmosis/issues/2208

## What is the purpose of the change

Adds multi-pool(balancer 3 asset pool) test for `AfterPoolCreated` hooks.
Most of the file differences come from renaming `PrepareUni2PoolWithAssets` to `PrepareUni2PoolWithCoins`, as I thought we had inconsistency with the naming conventions.

## Brief Changelog
- Adds tests for three asset pool  for the `AfterPoolCreated` hook

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? no
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? no
  - How is the feature or change documented? not documented

### CODE REVIEW

File: x/twap/listeners_test.go [hunk 1]
```diff
@@ -3,35 +3,99 @@ package twap_test
...
+
+			// check that creating a pool saved a new state of changed pools
+			s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+
+			// check that all twap records have been created for all denom pairs in pool
+			twapRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denomPairs0, _ := types.GetAllUniqueDenomPairs(denoms)
```
Comment (czarcas7ic) [inline x/twap/listeners_test.go:75]: What is the significance of the 0 in `denomPairs0`?

File: x/twap/listeners_test.go [hunk 2]
```diff
@@ -42,7 +106,7 @@ func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
 // TODO: Abstract this to be more table driven, and test more pool / block setups.
 func (s *TestSuite) TestSwapAndEndBlockTriggeringSave() {
```
Comment (czarcas7ic) [inline x/twap/listeners_test.go:107]: Do we benefit from testing both UniV2 and multi-asset pools here as well? (asking to learn, unsure if it would provide value)

**Review by czarcas7ic — APPROVED** [review]
Was able to follow the tests fairly easily, well done 🙂

File: app/apptesting/gamm.go [hunk 3]
```diff
@@ -14,22 +14,37 @@ var DefaultAcctFunds sdk.Coins = sdk.NewCoins(
 	sdk.NewCoin("baz", sdk.NewInt(10000000)),
 )
 
-// PrepareUni2PoolWithAssets returns a Univ2 pool with the initial liquidity being the provided balances.
-func (s *KeeperTestHelper) PrepareUni2PoolWithAssets(asset1, asset2 sdk.Coin) uint64 {
+// PrepareUni2PoolWithCoins returns a Univ2 pool with the initial liquidity being the provided balances.
+func (s *KeeperTestHelper) PrepareUni2PoolWithCoins(coin1, coin2 sdk.Coin) uint64 {
```
Comment (p0mvn) [inline app/apptesting/gamm.go:18]: Can we use `PrepareBalancerPoolWithCoins` with 2 coins instead?

Comment (mattverse) [reply app/apptesting/gamm.go:18]: Didn't know folks would be down for this change as well :) Im totally down to changing this

File: x/twap/listeners_test.go (hunk 1 above)
Comment (mattverse) [reply x/twap/listeners_test.go:75]: `GetAllUniqueDenomPairs` would return two slices of string, the first slice indicating the first asset of the pair, the second slice indicating the second asset of the pair. We would use both in normal cases, denomPairs0[1], denomParis1[1] or something similar

File: x/twap/listeners_test.go (hunk 2 above)
Comment (mattverse) [reply x/twap/listeners_test.go:107]: yes! These are WIP, I wanted to have them in separate PR so that the PR doesn't get too big

File: app/apptesting/gamm.go (hunk 3 above)
Comment (mattverse) [reply app/apptesting/gamm.go]: done!

File: x/twap/keeper_test.go [hunk 4]
```diff
@@ -14,11 +14,13 @@ import (
...
 var (
-	denom0            = "token/B"
-	denom1            = "token/A"
-	defaultUniV2Coins = sdk.NewCoins(sdk.NewInt64Coin(denom0, 1_000_000_000), sdk.NewInt64Coin(denom1, 1_000_000_000))
-	baseTime          = time.Unix(1257894000, 0).UTC()
-	tPlusOne          = baseTime.Add(time.Second)
+	denom0                 = "token/B"
+	denom1                 = "token/A"
+	denom2                 = "token/C"
+	defaultUniV2Coins      = sdk.NewCoins(sdk.NewInt64Coin(denom0, 1_000_000_000), sdk.NewInt64Coin(denom1, 1_000_000_000))
```
Comment (p0mvn) [inline x/twap/keeper_test.go:20]: nit: Consider renaming this to `defaultTwoAssetCoins` for a clearer distinction and relationship to  `defaultThreeAssetCoins`

File: x/twap/listeners_test.go [hunk 5]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+	three_asset_coins := defaultThreeAssetCoins
+
+	tests := []struct {
+		name      string
+		poolCoins sdk.Coins
+		// if this field is set true, we swap in the same block with pool creation
+		runSwap bool
+	}{
+		{
+			"Uni2 Pool, no swap on pool creation block",
```
Comment (p0mvn) [inline x/twap/listeners_test.go:25]: nit: for anyone unfamiliar with what univ2 means
```suggestion
			"two asset pool, no swap on pool creation block",
```

File: x/twap/listeners_test.go [hunk 6]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+			two_asset_coin,
+			false,
+		},
+		{
+			"Uni2 Pool, swap on pool creation block",
```
Comment (p0mvn) [inline x/twap/listeners_test.go:30]:
```suggestion
			"two asset pool, swap on pool creation block",
```

File: x/twap/listeners_test.go [hunk 7]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
-func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-	s.BeginNewBlock(false)
-	s.RunBasicSwap(poolId)
-
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
```
Comment (p0mvn) [inline x/twap/listeners_test.go:14]: This is named `TestAfterPoolCreatedHook` but I don't see a call to `AfterPoolCreated`

What exactly is being tested here?

File: x/twap/listeners_test.go [hunk 8]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
-
-// Tests that after a swap, we are triggering internal tracking logic for a pool.
-func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
-	poolId := s.PrepareUni2PoolWithAssets(defaultUniV2Coins[0], defaultUniV2Coins[1])
-	s.BeginNewBlock(false)
-	s.RunBasicSwap(poolId)
-
-	s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
```
Comment (p0mvn) [inline x/twap/listeners_test.go:12]:
```suggestion
// and the correct state entries have been created upon pool creation.
```

File: x/twap/listeners_test.go (hunk 7 above)
Comment (p0mvn) [reply x/twap/listeners_test.go:14]: I see - this is done inside `PrepareBalancerPoolWithCoins` that calls `CreatePool`.

Can we add a comment explaining that please?

File: x/twap/listeners_test.go [hunk 9]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	two_asset_coin := defaultUniV2Coins
+	three_asset_coins := defaultThreeAssetCoins
```
Comment (p0mvn) [inline x/twap/listeners_test.go:16]: On the earlier note, do we have a test that calls `afterCreatePool` directly and primarily focuses on testing just that? If not, can we please add it?

File: x/twap/listeners_test.go [hunk 10]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+			// check that creating a pool saved a new state of changed pools
+			s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
+
+			// check that all twap records have been created for all denom pairs in pool
+			twapRecords, err := s.twapkeeper.GetAllMostRecentRecordsForPool(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)
+			s.Require().NoError(err)
+			denomPairs0, _ := types.GetAllUniqueDenomPairs(denoms)
+			s.Require().Equal(len(denomPairs0), len(twapRecords))
```
Comment (p0mvn) [inline x/twap/listeners_test.go:70]: Can we make the same check on `denomPairs1` just in case?

File: x/twap/listeners_test.go [hunk 11]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	two_asset_coin := defaultUniV2Coins
+	three_asset_coins := defaultThreeAssetCoins
+
+	tests := []struct {
+		name      string
+		poolCoins sdk.Coins
+		// if this field is set true, we swap in the same block with pool creation
+		runSwap bool
```
Comment (p0mvn) [inline x/twap/listeners_test.go:22]: Does it make sense to have `expectedTwapRecordCouns` or `expectedTwapRecords` here so that we can have more specific assertions? 

Currently, I think that most assertions we make depend on another method in the internal logic, and if both break, we might get a false positive

File: app/apptesting/gamm.go [hunk 12]
```diff
@@ -14,20 +14,19 @@ var DefaultAcctFunds sdk.Coins = sdk.NewCoins(
...
-				Token:  asset2,
-			},
-		},
-	)
+// PrepareBalancerPoolWithCoins returns a balancer pool
+// consisted of given coins with equal weight.
+func (s *KeeperTestHelper) PrepareBalancerPoolWithCoins(coins ...sdk.Coin) uint64 {
+	var poolAssets []balancer.PoolAsset
+	for _, coin := range coins {
+		poolAsset := balancer.PoolAsset{
```
Comment (alexanderbez) [inline app/apptesting/gamm.go:22]: nit: no need to assign to a variable

File: x/twap/listeners_test.go [hunk 13]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
-
-	twap, err = s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), denom0, denom1)
-	s.Require().NoError(err)
-	s.Require().Equal(expectedTwap, twap)
-}
-
-// TODO: Write test for create pool, swap, and EndBlock in same block, we use post-swap spot price in record.
-
-// Tests that after a swap, we are triggering internal tracking logic for a pool.
-func (s *TestSuite) TestSwapTriggeringTrackPoolId() {
```
Comment (ValarDragon) [inline x/twap/listeners_test.go:29]: Why was this deleted?

Comment (ValarDragon) [reply x/twap/listeners_test.go:29]: we need tests for swaps against already existing pools

File: x/twap/listeners_test.go [hunk 14]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+			var poolId uint64
+			// prepare pool according to test case
+			poolId = s.PrepareBalancerPoolWithCoins(tc.poolCoins...)
+
+			if tc.runSwap {
+				s.RunBasicSwap(poolId)
+			}
+
+			timeBeforeBeginBlock := s.Ctx.BlockTime()
+			s.twapkeeper.EndBlock(s.Ctx)
```
Comment (ValarDragon) [inline x/twap/listeners_test.go:58]: why don't we run s.EndBlock() ? (or does no such method exist and we should write an issue?)

File: x/twap/listeners_test.go [hunk 15]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+				s.Require().Equal(asset0sp, twapRecord.P0LastSpotPrice)
+				s.Require().Equal(timeBeforeBeginBlock, twapRecord.Time)
+
+				// check that we have same state entry
+				storeRepresentationRecord, err := s.twapkeeper.GetMostRecentRecordStoreRepresentation(s.Ctx, poolId, twapRecord.Asset0Denom, twapRecord.Asset1Denom)
+				s.Require().Equal(twapRecord, storeRepresentationRecord)
+
+				twapRecordBeforeTime, err := s.twapkeeper.GetRecordAtOrBeforeTime(s.Ctx, poolId, s.Ctx.BlockTime(), twapRecord.Asset0Denom, twapRecord.Asset1Denom)
+				s.Require().NoError(err)
+				s.Require().Equal(twapRecord, twapRecordBeforeTime)
```
Comment (ValarDragon) [inline x/twap/listeners_test.go:89]: Can we just construct the expected TWAP record, and then check equality with one representation?

I don't think we need to check all representations, or if we do that should be a separate method - its a guarantee elsewhere in the code that all methods update together.

File: x/twap/listeners_test.go [hunk 16]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+			if tc.runSwap {
+				s.RunBasicSwap(poolId)
+			}
+
+			timeBeforeBeginBlock := s.Ctx.BlockTime()
+			s.twapkeeper.EndBlock(s.Ctx)
+			s.BeginNewBlock(false)
+
+			// check that creating a pool saved a new state of changed pools
+			s.Require().Equal([]uint64{poolId}, s.twapkeeper.GetChangedPools(s.Ctx))
```
Comment (ValarDragon) [inline x/twap/listeners_test.go:62]: err why didnt GetChangedPools get wiped on EndBlock?

File: x/twap/listeners_test.go [hunk 17]
```diff
@@ -3,35 +3,93 @@ package twap_test
...
+// TestAfterPoolCreatedHook tests if internal tracking logic has been triggered correctly,
+// and the corerct state entries have been created upon pool creation.
+// This test includes test cases for swapping on the same block with pool creation.
+func (s *TestSuite) TestAfterPoolCreatedHook() {
+	two_asset_coin := defaultUniV2Coins
+	three_asset_coins := defaultThreeAssetCoins
+
+	tests := []struct {
+		name      string
+		poolCoins sdk.Coins
```
Comment (ValarDragon) [inline x/twap/listeners_test.go:20]: We should be able to test multiple pool creations in one block

File: x/twap/listeners_test.go (hunk 16 above)
Comment (ValarDragon) [reply x/twap/listeners_test.go:62]: ahh were testing the flow wrong. Needs to have commit as done in `EndBlockTriggeringSave`.

This line should be moved up before EndBlock.

File: x/twap/listeners_test.go (hunk 10 above)
Comment (ValarDragon) [reply x/twap/listeners_test.go:70]: idk, I feel like were mixing what were testing. The goal is to just sanity check that we have the right number of records.

File: x/twap/listeners_test.go (hunk 11 above)
Comment (ValarDragon) [reply x/twap/listeners_test.go:18]: Good point. In the latest commit I made it get the correct expected record as `types.NewTwapRecord(s.App.GAMMKeeper, s.Ctx, poolId, denomPairs0[i], denomPairs1[i])`, which we then test equality against. (As that is our definition of correct here)

Only thing I think that would be slipping is if:
* `types.NewTwapRecord` returned something wrong for an integration test failure,
* 0 records were created, and our expected logic missed this due to a bug in
  *  `types.GetAllUniqueDenomPairs(denoms)`failing. 
  *  denoms, err := s.App.GAMMKeeper.GetPoolDenoms(s.Ctx, poolId)

File: x/twap/listeners_test.go (hunk 13 above)
Comment (ValarDragon) [reply x/twap/listeners_test.go:29]: reintroduced

File: x/twap/listeners_test.go (hunk 11 above)
Comment (ValarDragon) [reply x/twap/listeners_test.go:19]: I think trusting `types.NewTwapRecord` is fine, since thats our definition of correct for a new record. 

I also feel like `types.GetAllUniqueDenomPairs(denoms)` should be well tested / have no edge cases, but we could if we wanted add an 'expected num records' just to sanity check

**Review by ValarDragon — APPROVED** [review]
Committed a slight restructuring, to use the function we have to test if the expected record is correct.

Otherwise LGTM.

Only thing I think we don't check is what if multiple pools are created in a single block, do you think that should be added to this test, or be a separate one?

**Comment by mattverse** [comment]
Ty for the refactoring!

I think it'd be rather confusing to go around another iteration of reviews on this same PR, I'll be creating a new PR for that case
